File: linked_list/reverse_list/0234_palindrome_linked_list.py
# Given the head of a singly linked list, 
# return true if it is a palindrome or false otherwise.

# https://leetcode.com/problems/palindrome-linked-list/description/

class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if not head or not head.next: 
            return True
        l1 = head
        mid = self.middleNode(head)
        l2 = mid.next
        # cut the list into 2 halves
        mid.next = None
        # reverse 2nd half
        l2 = self.reverseList(l2)
        while l1 and l2:
            if l1.val != l2.val:
                return False
            l1 = l1.next
            l2 = l2.next
        return True
    # return first mid node for even length
    def middleNode(self, head):
        slow = fast = head
        while fast.next and fast.next.next: 
        # while fast and fast.next: -> return 2nd mid node
            slow = slow.next
            fast = fast.next.next
        return slow
    # reverse list
    def reverseList(self, head):
        pre = None
        cur = head
        while cur:
            nxt = cur.next
            cur.next = pre
            pre = cur
            cur = nxt
        return pre

File: linked_list/reverse_list/0234_palindrome_linked_list.js
/*
Given the head of a singly linked list, return true if it is a 
palindrome or false otherwise.
*/

// https://leetcode.com/problems/palindrome-linked-list/

// convert to array + two pointers
// time complexity O(n)
// space complexity O(n)
var isPalindrome = function (head) {
  let arr = [];
  while (head) {
    arr.push(head.val);
    head = head.next;
  }
  let left = 0,
    right = arr.length - 1;
  while (left < right) {
    if (arr[left] !== arr[right]) return false;
    left++;
    right--;
  }
  return true;
};

// reverse half
// time complexity O(n)
// space complexity O(n)
var isPalindrome = function (head) {
  var reverse = function (head) {
    let prev = null;
    let curr = head;
    while (curr) {
      const next = curr.next;
      [curr.next, prev, curr] = [prev, curr, next];
    }
    return prev;
  };
  if (!head || !head.next) return true;
  const dummyNode = new ListNode(0, head);
  let slow = dummyNode,
    fast = dummyNode;
  while (fast.next && fast.next.next) {
    slow = slow.next;
    fast = fast.next.next;
  }
  // cut the list into two halves: head -> slow, slow.next -> tail
  let secondHalf = fast.next ? slow.next.next : slow.next;
  slow.next = null;
  let reverseHead = reverse(head);
  while (secondHalf && reverseHead) {
    if (secondHalf.val !== reverseHead.val) return false;
    secondHalf = secondHalf.next;
    reverseHead = reverseHead.next;
  }
  return true;
};

// two pointers
var isPalindrome = function (head) {
  let slow = head,
    fast = head,
    prev = null;
  // reverse the first half
  while (fast && fast.next) {
    fast = fast.next.next;
    let next = slow.next;
    slow.next = prev;
    prev = slow;
    slow = next;
  }
  // adjust the slow pointer if the length is odd
  if (fast) slow = slow.next;
  // compare the first half and the second half
  while (prev) {
    if (prev.val !== slow.val) return false;
    slow = slow.next;
    prev = prev.next;
  }
  return true;
};
